### PR TITLE
Upgrade app service plans to Premium v3

### PR DESCRIPTION
**Pull request by nickclyde**
## Related Issue or Background Info

- After deploying #3312, we discovered an out of memory error. We are still looking into that issue, but we noticed that there is a new version of the app service plans, V3 ([released in July 2021](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/announcing-app-service-environment-v3-ga/ba-p/2517990)). By upgrading all of our app services to this version, we can potentially save money and also have faster CPUs and more memory. [Pricing page here](https://azure.microsoft.com/en-us/pricing/details/app-service/linux/#pricing)
## Changes Proposed

- Change all app service plans to V3
- Prod and stg use tier 2 `P3v3`, the rest use tier 1 `P1v3`

## Additional Information

- decisions that were made
- discussion of tradeoffs / future work
- complaints about how bad the code is

### Infrastructure
- [ ] **Consult the results of the `terraform-plan` job inside the "Terraform Checks" workflow run for this PR. Confirm that there are no unexpected changes!**

### Design
- [ ] Any UI/UX changes have a designer as a reviewer, and changes have been approved
- [ ] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke-tested by both the engineering and design teams

### Content
- [ ] Any content changes (including new error messages) have been approved by content team

### Support
- [ ] Any changes that might generate new support requests have been flagged to the support team
- [ ] Any changes to support infrastructure have been demo'd to support team

### Testing
- [ ] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [ ] Database changes are submitted as a separate PR
  - [ ] Any new tables that do not contain PII are accompanied by a GRANT SELECT to the no-PHI user
  - [ ] Any changes to tables that have custom no-PHI views are accompanied by changes to those views
        (including re-granting permission to the no-PHI user if need be)
  - [ ] Liquibase rollback has been tested locally using `./gradlew liquibaseRollbackSQL` or `liquibaseRollback`
  - [ ] Each new changeset has a corresponding [tag](https://docs.liquibase.com/change-types/community/tag-database.html)
- [ ] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed

## Cloud
- [ ] DevOps team has been notified if PR requires ops support
- [ ] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

### CODE REVIEW

**Review by rin-skylight — APPROVED** [review]
It looks like the right variables have been updated, `terraform plan` is reflecting everything correctly, and we have consensus on what to upgrade to. This looks great to me! Thanks for volunteering to take this on!

**Review by zdeveloper — COMMENTED** [review]
I don't have the expertise to say whether the tf changes look good or not, but it looks great to me otherwise! Thank you for doing this. @jdorothy @rin-skylight @alismx should we merge this after hours and put a maintenance banner with the with a maintenance window, I am not sure if there might be any extended down time from the IP address changes associated with changes to the app service

File: ops/prod/api.tf [hunk 1]
```diff
@@ -4,7 +4,7 @@ module "simple_report_api" {
   env    = local.env
 
   instance_count = 4
-  instance_size  = "P2v2"
+  instance_size  = "P3v3"
```
Comment (jdorothy) [inline ops/prod/api.tf:7]: The PR description says this should be `P2v3` - which value is correct? `P3v3` is a pretty big cost jump.

File: ops/stg/api.tf [hunk 2]
```diff
@@ -4,7 +4,7 @@ module "simple_report_api" {
   env    = local.env
 
   instance_count = 3
-  instance_size  = "P2v2"
+  instance_size  = "P3v3"
```
Comment (jdorothy) [inline ops/stg/api.tf:7]: See above

File: ops/prod/api.tf (hunk 1 above)
Comment (jdorothy) [reply ops/prod/api.tf:7]: Also, even if we just jump from v2 to v3 and keep the instance size the same, it's a ~50% cost increase.

Comment (nickclyde) [reply ops/prod/api.tf:7]: Whoops, sorry I forgot to update the PR description. I chatted with @zdeveloper about this this morning - apparently Tim Best approved this upgrade specifically. I'm not sure if we really need 32 GB of memory, but I think at least the 16 provided by P2v3 is needed given our increase in traffic

Comment (jdorothy) [reply ops/prod/api.tf:7]: We're all good then if that's the case.

**Review by jdorothy — APPROVED** [review]

**Comment by nickclyde** [comment]
@zdeveloper To answer your question, I don't believe this would cause any downtime? At least I can't find any reference to it in the [upgrade docs here](https://docs.microsoft.com/en-us/azure/app-service/app-service-configure-premium-tier). Please correct me if I'm wrong @jdorothy @rin-skylight

File: ops/prod/api.tf (hunk 1 above)
Comment (zdeveloper) [reply ops/prod/api.tf:7]: Yeah, after chatting with Tim Best, he said we can go to P3v3 and observe the audit log, and if its an over kill, its always easy to scale down to P2v3

**Comment by nickclyde** [comment]
Gotcha. I think we should probably merge this off hours then. I don't mind doing it around 9 or 10 PM PT, not sure when the best time would be

**Pull request merged**